### ISSUE TITLE
🐛 Bug: Google application is not verified

### ISSUE THREAD

**Issue by kierzniak** [keeweb/keeweb#2190]
### Priority

High

### Version: KeeWeb

v1.18.7

### Environment

Web

### Operating System

Mac

### Range of issue

The issue occurs on my own setup AND demo site

### Database Type

Issue occurs on both new and existing databases

### Environment Copy/Paste

```shell
KeeWeb v1.18.7 (dfc6201, 2021-07-18)
Environment: web
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36
```

### Description

I'm using KeeWeb without problems for couple of years but recently when I try to use KeeWeb with Google Drive after login I'm getting information that app/developer is not verified by Google.
<img width="712" alt="Zrzut ekranu 2024-07-23 o 13 48 05" src="https://github.com/user-attachments/assets/b0bc2588-76c7-49c6-a296-67db4670857a">
Translation
```
This app has not been reviewed by Google
This app asks for access to sensitive information about your Google account. We advise against using this app unless the developer (...@gmail.com) has verified it with Google.
```
Of course I can bypass this warning but I'm using KeeWeb in my company also and people are little scared because of this warning. 

### Steps To Reproduce

Using incognito mode in the browser and Google account which was never used with KeeWeb on particular machine. Or try on your smartphone.

1. Open https://app.keeweb.info/
2. Login to Google Drive
3. Error should be thrown

### Expected Behavior

Using incognito mode in the browser and Google account which was never used with KeeWeb on particular machine. Or try on your smartphone.

1. Open https://app.keeweb.info/
2. Login to Google Drive
3. Error should not be thrown

### Referenced Issues

- #

### Attachments: Screenshots

<img width="712" alt="Zrzut ekranu 2024-07-23 o 13 48 05" src="https://github.com/user-attachments/assets/2849e90a-baa9-4482-80d7-7530dc79310c">

Translation
```
This app has not been reviewed by Google
This app asks for access to sensitive information about your Google account. We advise against using this app unless the developer (...@gmail.com) has verified it with Google.
```

**Comment by maxiimilian** [comment]
I have the same problem with the native app on macOS. Unfortunately, it seems that the development of KeeWeb has stopped...

**Comment by Jurrie** [comment]
> Unfortunately, it seems that the development of KeeWeb has stopped...

Actually, this project has become active again recently! There are new maintainers. Currently, the codebase is brought back up to par. I don't know about MacOS support though. I think there was a lack of testers. Seeing as you use MacOS, maybe you could help out? If so, you should talk to Aetherinox.

**Comment by Aetherinox** [comment]
@Jurrie Correct.

I've got Windows, Linux, and Android covered. As of right now, I can't do fine testing on MacOS as I don't have one available. I've put it on my list of things. While most code will work across the board, it makes testing difficult without additional eyes. All I have to go on is reading through the code itself and seeing if something sticks out.

**Comment by antelle** [comment]
@Aetherinox I submitted a review for Google a while ago after they required a re-submission, but it was rejected because of Google Drive scope. Not sure what to do with that because passing the additional audit that they suggested seems complicated. It's also possible to remove the "all drive" scope, but then the Google Drive file selector won't work.
You can find the messages from Google in the google group to see the details.

**Comment by Aetherinox** [comment]
@antelle Thanks, I'll take a look. Maybe that's what confused me. I had a new developer app created, and I saw a message for an approval request which I hadn't applied for at the time as I was still working through some of the code.

Makes more sense now.

I'll go read the letter and see what the options are.

**Comment by antelle** [comment]
It's this message where I was stuck.

<img width="926" alt="image" src="https://github.com/user-attachments/assets/de17fbce-6bf8-45d3-b508-864949a8673c">

**Comment by kierzniak** [comment]
I don't have a experience with "Tier 3 assessment" but maybe "Drive API" will be sufficient?
https://developers.google.com/drive/api/guides/api-specific-auth#drive-scopes

Does using "Drive API" require a rewrite of the google drive implementation?

**Comment by maxiimilian** [comment]
> > Unfortunately, it seems that the development of KeeWeb has stopped...
> 
> Actually, this project has become active again recently! There are new maintainers. Currently, the codebase is brought back up to par. I don't know about MacOS support though. I think there was a lack of testers. Seeing as you use MacOS, maybe you could help out? If so, you should talk to Aetherinox.

Great news! I am happy to help testing!

**Comment by Aetherinox** [comment]
I haven't messed with CASA assessments, but I'm reading up on it now.

@maxiimilian If you have MacOS and would be willing to help test, I can keep you in mind when releases are about to roll. It would be nice to have a few, as it's not possible for a single person to catch every single bug; especially since there's so many different configuration types.

I'm looking at acquiring my own setup in the future so that I can also do my own testing which will help out dramatically. But for now, we can work something out so that the process is smooth.

**Comment by ion-xwood** [comment]
Hello everyone.  I have the same problem. I solved it this way.

Create google cloud API access
 - @see https://github.com/keeweb/keeweb/wiki/Dropbox-and-GDrive
 - @see https://console.cloud.google.com/apis/credentials?project=keeweb-434812
 - create project
 - oauth client ID (type webapp)
 - set auth redirect url: http://localhost:48149/oauth-result/gdrive.html
 - copy clientId
 - copy secretId
 - add test user to webapp (self email)
 - https://console.cloud.google.com/apis/credentials/consent
 - add google drive api to webapp
 - https://console.cloud.google.com/apis/dashboard
 
Configure keeweb
 - launch keeweb
 - settings
 - scroll to down
 - show advanced optiopns
 - go to dev console
 - run
 - kw.settings.set('gdrive', 'true');
 - kw.settings.set('gdriveClientId', '#goauth client ID#');
 - kw.settings.set('gdriveClientSecret', '#goauth client secret#');
 - check kw.settings.get();
 - exit app
 - run and check

**Comment by ChaunceyHoover** [comment]
Just in case anyone else runs into this before it's patched, #2191 is addressing the same issue and someone is actively working on a solution.

**Comment by sergey-koba-mobidev** [comment]
> Hello everyone. I have the same problem. I solved it this way.
> 
> Create google cloud API access
> 
> * @see https://github.com/keeweb/keeweb/wiki/Dropbox-and-GDrive
> * @see https://console.cloud.google.com/apis/credentials?project=keeweb-434812
> * create project
> * oauth client ID (type webapp)
> * set auth redirect url: http://localhost:48149/oauth-result/gdrive.html
> * copy clientId
> * copy secretId
> * add test user to webapp (self email)
> * https://console.cloud.google.com/apis/credentials/consent
> * add google drive api to webapp
> * https://console.cloud.google.com/apis/dashboard
> 
> Configure keeweb
> 
> * launch keeweb
> * settings
> * scroll to down
> * show advanced options
> * go to dev console
> * run
> * kw.settings.set('gdrive', 'true');
> * kw.settings.set('gdriveClientId', '#goauth client ID#');
> * kw.settings.set('gdriveClientSecret', '#goauth client secret#');
> * kw.settings.get();
> * exit app
> * run and check

I love you! This worked!

**Comment by Prototyped** [comment]
> Hello everyone. I have the same problem. I solved it this way.
> 
> Create google cloud API access
> 
> * @see https://github.com/keeweb/keeweb/wiki/Dropbox-and-GDrive
> * @see https://console.cloud.google.com/apis/credentials?project=keeweb-434812
> * create project
> * oauth client ID (type webapp)
> * set auth redirect url: http://localhost:48149/oauth-result/gdrive.html
> * copy clientId
> * copy secretId
> * add test user to webapp (self email)
> * https://console.cloud.google.com/apis/credentials/consent
> * add google drive api to webapp
> * https://console.cloud.google.com/apis/dashboard
> 
> Configure keeweb
> 
> * launch keeweb
> * settings
> * scroll to down
> * show advanced options
> * go to dev console
> * run
> * kw.settings.set('gdrive', 'true');
> * kw.settings.set('gdriveClientId', '#goauth client ID#');
> * kw.settings.set('gdriveClientSecret', '#goauth client secret#');
> * kw.settings.get();
> * exit app
> * run and check

This worked great on Windows. On Linux, the "Show dev tools" button does noting (somehow the Linux build lacks support for dev tools altogether). I worked around this by setting up remote inspection for the command:

```shell
keeweb --inspect=9229
```

then connecting to it from a Chromium browser (Chrome and Chromium: chrome://inspect; Microsoft Edge for Linux, which I used: edge://inspect)

The browser finds the Electron process (under "Remote Target", "electron/js2c/browser_init file:///"). Click "inspect". This brings up Dev Tools.

Unfortunately in the console we don't have access to the `kw` object, or any part of its object model, but we can use the Sources tab and the debugger to make the changes.

First, use the pause button on the toolbar. Then in the left sidebar with the source tree, go to file://, usr/share/keeweb-desktop/resources/app.asar, scripts, main.js.

Search in the source code for `function mainWindowFocus()` (which gets invoked each time you bring keyboard focus to the KeeWeb window, e.g. by clicking it).

Change it from

```javascript
function mainWindowFocus() {
    emitRemoteEvent('main-window-focus');
}
```

to

```javascript
function mainWindowFocus() {
    emitRemoteEvent('main-window-focus');
    appSettings['gdrive'] = 'true';
    appSettings['gdriveClientId'] = '<your-client-id-from-google-cloud-console>';
    appSettings['gdriveClientSecret'] = '<your-client-secret-from-google-cloud-console>';
    console.log('Saving!');
    console.log(appSettings);
    saveConfig('app-settings', JSON.stringify(appSettings));
}
```

Now click on the KeeWeb window, then in the DevTools window, resume execution by clicking the Play button (remember we paused execution earlier).

Look at the Console tab of the DevTools window, and confirm that you see "Saving!" and a JavaScript object (`appSettings`), which when you expand it, includes the `gdrive`, `gdriveClientId` and `gdriveClientSecret` settings you set above.

Now in my experience, Google auth still doesn't work. You need to close down KeeWeb and restart it, and the next time you should be able to go through Google auth.

Note that if you were previously using Google auth, you may need to go into Settings first and log out of Google Drive within the settings, then try to access Google Drive again. This time auth should go through if you set up the OAuth client correctly (including granting "drive" Data Access to it).

**Comment by maksym-arutyunyan** [comment]
> It's also possible to remove the "all drive" scope, but then the Google Drive file selector won't work.

Could the approach KeeWeb uses with Google Drive be adjusted so that it requires users to create a dedicated KeeWeb folder on Google Drive and limits access to only that folder? 

I’d like KeeWeb to work with my files on Google Drive, and I’m perfectly fine with it being restricted to a single folder.

**Comment by SaadOps** [comment]
A simple workaround would be simply installing Drive Desktop and then while saving choose the drive folder.

**Comment by douglasdnzcarvalho** [comment]
It's interesting to have the workarounds and such, but what I liked most was being able to access both on desktop and mobile, so, any hope of one day finding a real solution to the problem?

**Comment by Aetherinox** [comment]
@douglasdnzcarvalho Aside from waiting on Google to finish completely our review, and us dying of old age, I've branched the submitted app, and I'm going to be utilizing the other recommendation which is to integrate https://developers.google.com/workspace/drive/api/guides/picker.

If Google ever finishes, then we can have multiple ways to access files, but until then, I feel like I'm going to celebrate retirement before Google gets around to finishing our app process. I get that they may be busy, and we can do nothing about that, but in our instance, that scope is vital to saving.

And then I need to also match the Microsoft end-point as well since that has been going haywire for some users.

Another user also asked if there could be an in-app way of specifying API keys for users who want to use the full-scope Google permissions, so I'm going to throw in some API tokens and wrap some storage security around it.
